mypyc/genops_for.py

```diff
@@ -0,0 +1,307 @@
+        builder.assign(self.index_target, builder.primitive_op(
+            unsafe_short_add,
+            [builder.read(self.index_target, line),
+             builder.add(LoadInt(1))], line), line)
+
+
+class ForRange(ForGenerator):
+    """Generate optimized IR for a for loop over an integer range."""
+
+    # TODO: Use a separate register for the index to allow safe index mutation.
+
+    def init(self, start_reg: Value, end_reg: Value) -> None:
+        builder = self.builder
+        self.start_reg = start_reg
+        self.end_reg = end_reg
+        self.end_target = builder.maybe_spill(end_reg)
+        # Initialize loop index to 0. Assert that the index target is assignable.
+        self.index_target = builder.get_assignment_target(
+            self.index)  # type: Union[Register, AssignmentTarget]
+        builder.assign(self.index_target, start_reg, self.line)
+
+    def gen_condition(self) -> None:
+        builder = self.builder
+        line = self.line
+        # Add loop condition check.
+        comparison = builder.binary_op(builder.read(self.index_target, line),
+                                       builder.read(self.end_target, line), '<', line)
+        builder.add_bool_branch(comparison, self.body_block, self.loop_exit)
+
+    def gen_step(self) -> None:
+        builder = self.builder
+        line = self.line
+
+        # Increment index register. If the range is known to fit in short ints, use
+        # short ints.
+        if (is_short_int_rprimitive(self.start_reg.type)
+                and is_short_int_rprimitive(self.end_reg.type)):
+            new_val = builder.primitive_op(
+                unsafe_short_add, [builder.read(self.index_target, line),
+                                   builder.add(LoadInt(1))], line)
+        else:
+            new_val = builder.binary_op(
+                builder.read(self.index_target, line), builder.add(LoadInt(1)), '+', line)
+        builder.assign(self.index_target, new_val, line)
+
+
+class ForInfiniteCounter(ForGenerator):
+    """Generate optimized IR for a for loop counting from 0 to infinity."""
+
+    def init(self) -> None:
+        builder = self.builder
+        # Initialize loop index to 0.
+        self.index_target = builder.get_assignment_target(
+            self.index)  # type: Union[Register, AssignmentTarget]
+        builder.assign(self.index_target, builder.add(LoadInt(0)), self.line)
+
+    def gen_step(self) -> None:
+        builder = self.builder
+        line = self.line
+        # We can safely assume that the integer is short, since we are not going to wrap
+        # around a 63-bit integer.
+        # NOTE: This would be questionable if short ints could be 32 bits.
+        new_val = builder.primitive_op(
+            unsafe_short_add, [builder.read(self.index_target, line),
+                               builder.add(LoadInt(1))], line)
+        builder.assign(self.index_target, new_val, line)
+
+
+class ForEnumerate(ForGenerator):
```

Comment:
A future optimization might be special casing enumerate on lists to reuse the index counter